### PR TITLE
[feat]: Add Image-Video Mixture training to main repo

### PR DESCRIPTION
**Pull request by rlsu9**

### CODE REVIEW

File: fastvideo/dataset/latent_datasets.py [hunk 1]
```diff
@@ -22,18 +22,18 @@ def __init__(
         self.prompt_attention_mask_dir = os.path.join(self.datase_dir_path, "prompt_attention_mask")
         with open(self.json_path, 'r') as f:
             self.data_anno = json.load(f)
-        self.data_anno = sorted(self.data_anno, key=lambda x: x['latent_path'])
+        # json.load(f) already keeps the order
+        # self.data_anno = sorted(self.data_anno, key=lambda x: x['latent_path'])
         self.num_latent_t = num_latent_t
         self.uncond_prompt_embed = torch.load(os.path.join(uncond_prompt_embed_mask_dir, "embed.pt"), map_location="cpu", weights_only=True)
         self.uncond_prompt_mask = torch.load(os.path.join(uncond_prompt_embed_mask_dir, "mask.pt"), map_location="cpu", weights_only=True)
+        self.lengths = [torch.load(os.path.join(self.latent_dir, data_item["latent_path"])).shape[1] for data_item in self.data_anno]
```
Comment (jzhang38) [inline fastvideo/dataset/latent_datasets.py]: Very inefficient.

File: fastvideo/dataset/t2v_datasets.py [hunk 2]
```diff
@@ -182,12 +181,15 @@ def get_image(self, idx):
...
+        image = image.unsqueeze(0)
+        
+        image = image.float() / 127.5 - 1.0
+        
         caps = image_data['cap'] if isinstance(image_data['cap'], list) else [image_data['cap']]
         caps = [random.choice(caps)]
         text = text_preprocessing(caps, support_Chinese=self.support_Chinese)
         input_ids, cond_mask = [], []
-        text = text if random.random() > self.cfg else ""
+        text = text
```
Comment (jzhang38) [inline fastvideo/dataset/t2v_datasets.py]: Why delete this?

File: fastvideo/dataset/latent_datasets.py [hunk 3]
```diff
@@ -22,18 +22,18 @@ def __init__(
         self.prompt_attention_mask_dir = os.path.join(self.datase_dir_path, "prompt_attention_mask")
         with open(self.json_path, 'r') as f:
             self.data_anno = json.load(f)
-        self.data_anno = sorted(self.data_anno, key=lambda x: x['latent_path'])
+        # json.load(f) already keeps the order
+        # self.data_anno = sorted(self.data_anno, key=lambda x: x['latent_path'])
```
Comment (jzhang38) [inline fastvideo/dataset/latent_datasets.py:26]: Why comment here?

File: fastvideo/utils/dataset_utils.py [hunk 4]
```diff
@@ -326,4 +299,14 @@ def __len__(self):
...
+        def get_sublist(lst, rank, batch_size, world_size):
+            result = []
+            index = rank * batch_size
+            while index < len(lst):
+                result.extend(lst[index:index + batch_size])
+                index += batch_size * world_size
+            return result
+        
+        indices = get_sublist(indices, self.rank, self.batch_size, self.world_size)
+        print(f"rank {self.rank} gots", indices[:20] )
```
Comment (jzhang38) [inline fastvideo/utils/dataset_utils.py]: what is this

File: fastvideo/dataset/latent_datasets.py (hunk 1 above)
Comment (rlsu9) [reply fastvideo/dataset/latent_datasets.py]: need to be fix with data preprocess

**Comment by ghost** [comment]
import json
import os
import shutil
import subprocess

meta_path = "/ephemeral/hao.zhang/codefolder/FastVideo-OSP/data/inital_training_dataset/overall"
meta_json_data_path = "/ephemeral/hao.zhang/codefolder/FastVideo-OSP/data/inital_training_dataset/overall.json"
save_path = "/ephemeral/hao.zhang/codefolder/FastVideo-OSP/data/inital_training_dataset_reformat"

json_data = None
with open(meta_json_data_path, "r") as f:
    json_data = json.load(f)
    
video_name_list = json_data["captions"].keys()

os.makedirs(os.path.join(save_path, "videos"), exist_ok=True)
target_json_path = "/ephemeral/hao.zhang/codefolder/FastVideo-OSP/data/inital_training_dataset_reformat/video2caption.json"

new_json_data = []
for idx, video_name in enumerate(video_name_list):
    new_name = str(idx) + ".mp4"
    source_path = os.path.join(meta_path, video_name)
    target_path = os.path.join(save_path, "videos", new_name)
    shutil.copy(source_path, target_path)
    ffprobe_cmd = [
            "ffprobe",
            "-v", "error",
            "-select_streams", "v:0",
            "-show_entries", "stream=width,height,r_frame_rate",
            "-show_format",
            "-of", "json",
            target_path
        ]
    result = subprocess.run(ffprobe_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
    metadata = json.loads(result.stdout)

    # Extract width, height, fps, and duration
    video_stream = metadata.get("streams", [])[0]
    width = video_stream.get("width")
    height = video_stream.get("height")
    
    # FPS calculation
    r_frame_rate = video_stream.get("r_frame_rate", "0/0")
    fps = eval(r_frame_rate) if "/" in r_frame_rate else float(r_frame_rate)
    
    # Duration (from format section)
    duration = float(metadata["format"]["duration"])
    
    # File size in bytes
    size = os.path.getsize(target_path)

    # Create the final dictionary
    video_info = {
        "path": new_name,
        "resolution": {
            "width": width,
            "height": height
        },
        "size": size,
        "cap": json_data["captions"][video_name]["caption"],
        "fps": round(fps, 2),
        "duration": round(duration, 2)
    }
    new_json_data.append(video_info)
    
with open(target_json_path, "w") as f:
    json.dump(new_json_data, f, indent=4)

**Comment by ghost** [comment]
https://github.com/user-attachments/assets/b2b6dc6d-50fb-4eef-bbaf-56c615015b98

**Comment by ghost** [comment]
Our machine provider and I are ready to replace the misbehaving machine - the root disk will be migrated automatically. Let me know if there's any reason I need to wait, otherwise I will give them the go-ahead for replacement on Monday.
The operations should take ~1 day or less.
Thanks, and have a good weekend!

**Comment by ghost** [comment]
export WANDB_MODE=online
torchrun --nnodes 1 --nproc_per_node 6 \
    fastvideo/train.py \
    --seed 42 \
    --pretrained_model_name_or_path data/mochi \
    --cache_dir "data/.cache" \
    --data_json_path "data/Image-Train-Full/videos2caption.json" \
    --validation_prompt_dir "data/Encoder_Overfit_Data/validation_prompt_embed_mask" \
    --uncond_prompt_dir "data/Encoder_Overfit_Data/uncond_prompt_embed_mask" \
    --gradient_checkpointing \
    --train_batch_size=32 \
    --num_latent_t 1 \
    --sp_size 1 \
    --train_sp_batch_size 1 \
    --dataloader_num_workers 4 \
    --gradient_accumulation_steps=2 \
    --max_train_steps=20000 \
    --learning_rate=1e-5 \
    --mixed_precision="bf16" \
    --checkpointing_steps=250 \
    --validation_steps 50 \
    --validation_sampling_steps 64 \
    --checkpoints_total_limit 3 \
    --allow_tf32 \
    --ema_start_step 0 \
    --cfg 0.0 \
    --ema_decay 0.999 \
    --log_validation \
    --output_dir="data/outputs/Image_Train" \
    --weighting_scheme "uniform"
